[kintatho] iP

KinTatHo opened the pull request:
# Yoda Java Chatbot

## Introduction
Meet Yoda, not the Jedi Master, but your personal task-master! This friendly Java chatbot, inspired by the wisdom of its namesake, helps you manage your tasks with Yoda-like sagacity. Through a simple command-line interface, Yoda guides you in adding, deleting, marking, and listing tasks with a touch of intergalactic fun.

## Galactic Features

### Task Mastery
- **Add Tasks**: Summon different types of tasks — Todos, Deadlines, and Events — with a mere command.
- **Delete Tasks**: Banish unwanted tasks from your list with the might of a Jedi.
- **Mark Tasks**: Declare tasks as conquered (done) or pending (undone).
- **List Tasks**: Display your tasks in an organized list, as if viewing a map of the galaxy.

### The Force of Data Persistence
- **Save/Load**: Fear not, young Padawan, for your tasks are automatically saved and retrieved, just like the ancient Jedi texts.

## Usage

### `todo` - Add a Todo Task
Command Yoda to add a Todo task to your list.

Example of usage: 
```java
todo read the Jedi archives
```
Expected outcome:
Hmm, added this task, I have:
[T][ ] read the Jedi archives
Tasks in the list, now you have 1, hmm.

### `deadline` - Add a Deadline Task
Instruct Yoda to remember a task with a specific due date.

Example of usage:
```java
deadline build a new lightsaber /by 2024-05-05
```

Expected outcome:
Hmm, added this task, I have:
[D][ ] build a new lightsaber (by: May 05 2024)
Tasks in the list, now you have 2, hmm.

### `event` - Add an Event Task
Tell Yoda to note down an event with a start and end time.

Example of usage:
```java
event Rebel Alliance meeting /from 2024-05-05 14:00 /to 2024-05-05 16:00
```

Expected outcome:
Hmm, added this task, I have:
[E][ ] Rebel Alliance meeting (from: May 05 2024 14:00 to May 05 2024 16:00)
Tasks in the list, now you have 3, hmm.

### `list` - List All Tasks
Ask Yoda to reveal all your current tasks.

Example of usage:
```java
list
```

Expected outcome:
Tasks in your list, here they are:
1. [T][ ] read the Jedi archives
2. [D][ ] build a new lightsaber (by: May 05 2024)
3. [E][ ] Rebel Alliance meeting (from: May 05 2024 14:00 to May 05 2024 16:00)

MORE COMING SOON:

- [ ] Finding HAN SOLO

## 🛠 Installation
Click [here](https://github.com/KinTatHo/ip/releases/tag/A-Jar) to download Yoda on your system.
> "Do or do not. There is no try." - Yoda

 
**MAY THE FORCE** be with you in your task management journey! 👋👋

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

KinTatHo committed:
Level-5

KinTatHo committed:
A-Enums

KinTatHo committed:
Level 6

KinTatHo committed:
Level-5

KinTatHo committed:
Level-5

KinTatHo committed:
This is Level-5

KinTatHo committed:
Added Level-7

KinTatHo committed:
This is level-7

KinTatHo committed:
merge conflicts

KinTatHo committed:
Resolved Deadline merge conflicts

KinTatHo committed:
changed to deadline

KinTatHo committed:
Created Parser class, DateTimeUtil, and Level-8

KinTatHo committed:
Merge branch 'branch-level-8'
yes

KinTatHo committed:
A-MoreOOP

KinTatHo committed:
saving method changed to reading the file first, adding packages

KinTatHo committed:
Merge conflicts done i hope

KinTatHo committed:
added workspace.xml

KinTatHo committed:
commited staged files

KinTatHo committed:
merge conflicts

KinTatHo committed:
added test cases for TaskList, Todo

KinTatHo committed:
Added JavaDocs for Storage

KinTatHo committed:
Merge branch 'branch-A-JavaDoc'

KinTatHo committed:
Changed build.gradle mainclass, added TaskListTest and TodoTest

KinTatHo committed:
changed imports for storage class from *

KinTatHo committed:
Added findTasks method in TaskList, made changes to Command, Parser and YodaUI to fit findTasks

KinTatHo committed:
Revert "Added findTasks method in TaskList, made changes to Command, Parser and YodaUI to fit findTasks"

This reverts commit ec6042b8c5dc7fba8ec52d0e857b31cd41fcc4eb.

KinTatHo committed:
TaskList.java: added findTasks method, YodaUI: added findTasks method, Command.java: added FIND enum, Parser.java: added FIND case

KinTatHo committed:
TaskList.java: added comments for methods

KinTatHo committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/yoda/task/TaskList.java

macareonie commented on src/main/java/Yoda.java:
```diff
@@ -0,0 +1,51 @@
+ * Yoda can also find tasks by searching for keywords.
+ */
+public class Yoda {
+    private final YodaUI yodaUI;
```
```suggestion
    private final YodaUI YODA_UI;
```

constants should be uppercase with underscore to separate words.

macareonie commented on src/main/java/yoda/YodaUI.java:
```diff
@@ -0,0 +1,158 @@
+    private final TaskList taskList;
+    private final String chatbotName;
+    private boolean isChatting;
+    private final Storage storage;
```
```suggestion
    private final TaskList TASKLIST;
    private final String CHATBOT_NAME;
    private boolean isChatting;
    private final Storage STORAGE;
```

constants should be uppercase with underscore to separate words.

macareonie commented on src/main/java/yoda/components/Replies.java:
```diff
@@ -0,0 +1,39 @@
+package yoda.components;
+
+public class Replies {
+    private final String chatbotName;
```
```suggestion
    private final String CHATBOT_NAME;
```
constants should be uppercase with underscore to separate words.

macareonie commented on src/main/java/Yoda.java:
```diff
@@ -0,0 +1,51 @@
+import yoda.YodaUI;
```
Could possibly create a yoda package and put the existing code into it as per the A-Packages minimal requirement

macareonie commented on src/main/java/yoda/parser/Parser.java:
```diff
@@ -0,0 +1,152 @@
+                    break;
+                case EVENT:
+                    addTaskWithDateTime(parts, Command.EVENT);
+                    break;
```
```suggestion
            switch (command) {
            case BYE:
                yodaUI.printMessage("Farewell. See you again, I hope!");
                yodaUI.stopChatting();
                break;
            case LIST:
                yodaUI.showTasks();
                break;
            case SAVE:
                TaskList taskList = yodaUI.getTaskList();
                yodaUI.saveTasks(taskList);
                break;
            case DELETE:
                performTaskOperation(parts, yodaUI::deleteTask);
                break;
            case FIND:
                if (parts.length > 1) {
                    yodaUI.findTasks(parts[1]);
                } else {
                    throw new Exception("Search term, provide you must.");
                }
                break;
            case MARK:
                performTaskOperation(parts, yodaUI::markTaskAsDone);
                break;
            case UNMARK:
                performTaskOperation(parts, yodaUI::markTaskAsUndone);
                break;
            case TODO:
                yodaUI.addTask(new Todo(parts[1]));
                break;
            case DEADLINE:
                addTaskWithDateTime(parts, Command.DEADLINE);
                break;
            case EVENT:
                addTaskWithDateTime(parts, Command.EVENT);
                break;
```

The switch statement should not have indentation for case clauses. Can configure IDE settings to make things easier.

macareonie commented on src/main/java/yoda/storage/Storage.java:
```diff
@@ -0,0 +1,203 @@
+        String description = parts[2];
+
+        switch (type) {
+            case "T":
+                return createTodoTask(isDone, description);
+            case "D":
+                return createDeadlineTask(isDone, description, parts[3]);
+            case "E":
+                return createEventTask(isDone, description, parts[3], parts[4]);
+            default:
```
switch statement: same as above

macareonie commented on src/main/java/yoda/storage/Storage.java:
```diff
@@ -0,0 +1,203 @@
+
+/** Represents the storage for tasks. */
+public class Storage {
+    private final String filePath;
+    private final Path path;
```
constant variable: same as above

macareonie commented on src/main/java/yoda/parser/Parser.java:
```diff
@@ -0,0 +1,152 @@
+import yoda.datetimeutil.DateTimeUtil;
+public class Parser {
+
+    private final YodaUI yodaUI;
```
same as above

macareonie commented on src/main/java/yoda/task/Deadline.java:
```diff
@@ -0,0 +1,53 @@
+ * Represents a deadline task. It extends the Task class and adds a deadline time.
+ */
+public class Deadline extends Task {
+    private final LocalDateTime by; // Deadline time for the task
```
constant variable: same as above

macareonie commented on src/main/java/yoda/task/Event.java:
```diff
@@ -0,0 +1,72 @@
+ * Represents an event task.
+ */
+public class Event extends Task {
+    private final LocalDateTime from; // Event starting time
+    private final LocalDateTime to; // Event ending time
```
constant variable: same as above

macareonie commented on src/main/java/yoda/task/Task.java:
```diff
@@ -0,0 +1,62 @@
+ */
+public class Task {
+
+    private final String description; // Description of the task
```
constant variable: same as above

macareonie commented on src/main/java/yoda/task/TaskList.java:
```diff
@@ -0,0 +1,146 @@
+ * Represents a list of tasks. It provides operations to add, delete, and retrieve tasks.
+ */
+public class TaskList {
+    private final List<Task> tasks; // List to store the tasks
```
constant variable: same as above

macareonie reviewed (commented):
Other than uppercasing constant variables and removing indentation for switch statement, overall coding standard has been adhered to. Possible nitpick would be ensuring the import statements are consistent through out the classes, such as whether there is a line break in between different dependencies or not.

Overall, really nice and clean code to read! 👍

KinTatHo committed:
changes constants into screaming snake case, changes switch indentations

KinTatHo committed:
Merge branch 'branch-Level-9'

Ma-Yirui reviewed (commented):
Great job keeping all the method names in line with the naming conventions!
Properly distinguished actions and things using verbs and nouns.
Methods names are easy to understand.

Keep up the good work!

KinTatHo committed:
adds GUI before refactoring

KinTatHo committed:
creates a GUI for yoda

KinTatHo committed:
adds comments to DialogBox.java

KinTatHo committed:
adds comments to Main.java

KinTatHo committed:
Merge branch 'branch-Level-10'

KinTatHo committed:
Create new Exceptions classes to differentiate exceptions.
Remove Chatbot Name variable in YodaUI class, chatbot name is Yoda throughout
Task attribute classes (e.g Todo, Deadline, Event) have some common behaviours (e.g markAsDone(), getStatus())
Convert Task into an abstract class

The common behaviors across person attribute classes cause code duplication.

Create a Replies class under constants package to replace constant strings in Yoda.java

KinTatHo committed:
Merge branch 'refactor-code'
Create new Exceptions classes to differentiate exceptions.
Remove Chatbot Name variable in YodaUI class, chatbot name is Yoda throughout
Task attribute classes (e.g Todo, Deadline, Event) have some common behaviours (e.g markAsDone(), getStatus())
Convert Task into an abstract class

The common behaviors across person attribute classes cause code duplication.

Create a Replies class under constants package to replace constant strings in Yoda.java

KinTatHo committed:
Add assertions to Parser.java

KinTatHo committed:
Merge branch 'branch-A-Assertions'
Merge branch-A-Assertions: add assertions to Parser class

KinTatHo committed:
Refactor code of Parser, Storage, TaskList and YodaUI classes

KinTatHo committed:
Merge pull request #1 from KinTatHo/branch-A-CodeQuality

Refactor code of Parser, Storage, TaskList and YodaUI classes

KinTatHo committed:
Add startUpdateProcess in YodaUI and helper methods

KinTatHo committed:
Yoda can now update properly

KinTatHo committed:
Merge branch 'branch-BCD-Extension'
Merge Branch-BCD-Extension

KinTatHo committed:
Add comments to new methods in YodaUI

KinTatHo committed:
Merge branch 'Branch-BCD-Extension'
Merge Branch-BCD-Extension

KinTatHo committed:
Merge pull request #2 from KinTatHo/branch-BCD-Extension

Add comments to new methods in YodaUI

KinTatHo committed:
Merge branch 'master' of https://github.com/KinTatHo/ip
Merge branch-BCD-Extension

KinTatHo committed:
Change yodaUI package name to ui as per naming conventions

KinTatHo committed:
Add a Ui.png based on the Yoda chatbot GUI. Update README.md to include new features

KinTatHo committed:
Merge pull request #4 from KinTatHo/userguide

Add Ui.png, update README.md

KinTatHo committed:
Add more tests using AI, add Ui.png

KinTatHo committed:
Merge pull request #5 from KinTatHo/Ui.png

Add more tests using AI, add Ui.png

KinTatHo committed:
Update the UserGuide

KinTatHo committed:
Merge pull request #6 from KinTatHo/Published-UG

Update the UserGuide

KinTatHo committed:
Re-add Ui.png file, debugged method for Storage

KinTatHo committed:
Merge pull request #7 from KinTatHo/Ui.png

Re-add Ui.png file, debugged method for Storage
